server/src/main/java/org/opensearch/repositories/RepositoriesService.java

```diff
@@ -210,30 +211,41 @@ public ClusterState execute(ClusterState currentState) {
                     RepositoriesMetadata repositories = metadata.custom(RepositoriesMetadata.TYPE);
                     if (repositories == null) {
                         logger.info("put repository [{}]", request.name());
+                        ensureNonSystemRepository(request.settings(), request.name());
                         repositories = new RepositoriesMetadata(
                             Collections.singletonList(new RepositoryMetadata(request.name(), request.type(), request.settings()))
                         );
                     } else {
                         boolean found = false;
                         List<RepositoryMetadata> repositoriesMetadata = new ArrayList<>(repositories.repositories().size() + 1);
-
+                        RepositoryMetadata currentRepositoryMetadata = null;
                         for (RepositoryMetadata repositoryMetadata : repositories.repositories()) {
                             if (repositoryMetadata.name().equals(newRepositoryMetadata.name())) {
                                 if (newRepositoryMetadata.equalsIgnoreGenerations(repositoryMetadata)) {
                                     // Previous version is the same as this one no update is needed.
                                     return currentState;
+                                } else if (SYSTEM_REPOSITORY_SETTING.get(repositoryMetadata.settings()) != SYSTEM_REPOSITORY_SETTING.get(
+                                    newRepositoryMetadata.settings()
+                                )) {
+                                    throw new RepositoryException(
+                                        repositoryMetadata.name(),
+                                        "trying to modify system repository attribute for a repository"
+                                    );
                                 }
                                 found = true;
+                                currentRepositoryMetadata = repositoryMetadata;
                                 repositoriesMetadata.add(newRepositoryMetadata);
                             } else {
                                 repositoriesMetadata.add(repositoryMetadata);
                             }
                         }
                         if (!found) {
                             logger.info("put repository [{}]", request.name());
+                            ensureNonSystemRepository(newRepositoryMetadata.settings(), newRepositoryMetadata.name());
```

Comment:
can we move this earlier where we also do `validateRepositoryMetadataSettings` ?